### PR TITLE
Collations 1.3

### PR DESCRIPTION
**Pull request by vmg**
## Description

More collations work! I'll add a few more patches to this between today and tomorrow.

- @frouioui requested this functionality: https://github.com/vitessio/vitess/commit/1c665d8896f40e981791af9b1ed837755abb3ca9

## Related Issue(s)
- https://github.com/vitessio/vitess/issues/8606

## Checklist
- [ ] Should this PR be backported?
- [ ] Tests were added or are not required
- [ ] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/collations/env.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	all = make([]Collation, 0, len(env.byID))
+	for _, col := range env.byID {
+		col.Init()
+		all = append(all, col)
+	}
+	return
+}
+
+// NewEnvironment creates a collation Environment for the given MySQL Version
+func NewEnvironment(version Version) *Environment {
```
Comment (frouioui) [inline go/mysql/collations/env.go:103]: Can we have a global `*Environment` here that gets created when calling this method or another one, so we don't have to keep a `*Environment` in the caller, we can just use `collations.GlobalEnvironment`, i.e:

```go
// instantiate the global environment
collations.NewEnvironment(MySQL5.7)

// use the global environment without having to store it
collations.GlobalEnvironment.DefaultCollationForCharset("utf8")
```

File: go/mysql/collations/internal/codegen/codegen.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package codegen
+
```
Comment (frouioui) [inline go/mysql/collations/internal/codegen/codegen.go:2]: header is missing

File: go/mysql/collations/tools/makemysqlversions/makemysqlversions.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
+package main
+
+import (
```
Comment (frouioui) [inline go/mysql/collations/tools/makemysqlversions/makemysqlversions.go:3]: missing header too

**Review by frouioui — APPROVED** [review]

File: go/mysql/collations/env.go (hunk 1 above)
Comment (vmg) [reply go/mysql/collations/env.go]: Yikes, that's extremely nasty API design. That's not happening, sorry. If you need a global collation environment, you can keep it on the `gen4` planner package... But I suspect @systay will also scream a you a little bit. :)

Spend some time thinking more carefully where the environment needs to live. I'm confident "globally for the whole process" is not the right answer.

**Pull request merged**